#### Задание 1.
Импортируйте библиотеки pandas и numpy.
Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn. Создайте датафреймы X и y из этих данных.
Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так, чтобы размер тестовой выборки
составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42.
Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.
Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.
Вычислите R2 полученных предказаний с помощью r2_score из модуля sklearn.metrics.


In [2]:
import warnings
warnings.filterwarnings('ignore')

In [3]:
import numpy as np
import pandas as pd

In [4]:
from sklearn.datasets import load_boston

In [5]:
boston = load_boston()
boston.keys()

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename'])

In [15]:
data = boston.data

In [16]:
data.shape

(506, 13)

In [17]:
feature_names = boston['feature_names']
feature_names

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

In [18]:
target = boston.target
target[:5]

array([24. , 21.6, 34.7, 33.4, 36.2])

In [19]:
#создаем два датафрэйма
X = pd.DataFrame(data, columns=feature_names)
X.head(3)

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03

In [20]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 13 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    float64
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    float64
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
dtypes: float64(13)
memory usage: 51.5 KB


In [21]:
def reduce_mem_usage(X):
    start_mem = X.memory_usage().sum() / 1024**2
    print('Memory usage of dataframe is {:.2f} MB'.format(start_mem))
    
    for col in X.columns:
        col_type = X[col].dtype
        if col_type !=object:
            c_min = X[col].min()
            c_max = X[col].max()
            if str(col_type)[:3] == 'int':
                if c_min > np.iinfo(np.int8).min and c_max < np.iinfo(np.int8).max:
                    X[col] = X[col].astype(np.int8)
                elif c_min > np.iinfo(np.int16).min and c_max < np.iinfo(np.int16).max:
                    X[col] = X[col].astype(np.int16)
                elif c_min > np.iinfo(np.int32).min and c_max < np.iinfo(np.int32).max:
                    X[col] = X[col].astype(np.int32)
                elif c_min > np.iinfo(np.int64).min and c_max < np.iinfo(np.int64).max:
                    X[col] = X[col].astype(np.int64)
            else:
                if c_min > np.finfo(np.float32).min and c_max < np.finfo(np.float32).max:
                    X[col] = X[col].astype(np.float32)
                else:
                    X[col] = X[col].astype(np.float64)
        else:
            X[col] = X[col].astype('category')
    end_mem = X.memory_usage().sum() / 1024**2
    print ('Memory usage after optimization is: {:.2f}MB'.format(end_mem))
    print('Decreased by {:.1f}%'.format(100*(start_mem - end_mem)/start_mem))
    return X     
X = reduce_mem_usage(X)

Memory usage of dataframe is 0.05 MB
Memory usage after optimization is: 0.03MB
Decreased by 49.9%


In [22]:
y = pd.DataFrame(target, columns=['price'])
y.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   price   506 non-null    float64
dtypes: float64(1)
memory usage: 4.1 KB


In [23]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [24]:
from sklearn.linear_model import LinearRegression

In [25]:
lr = LinearRegression()

In [26]:
lr.fit(X_train, y_train)

LinearRegression()

In [27]:
y_pred = lr.predict(X_test)
y_pred.shape

(152, 1)

In [28]:
check_test = pd.DataFrame({'y_test': y_test['price'], 'y_pred': y_pred.flatten()},
                          columns = ['y_test', 'y_pred'])
check_test.head()

y_test     y_pred
173    23.6  28.648939
274    32.4  36.495049
491    13.6  15.411057
72     22.8  25.403200
452    16.1  18.855175

In [29]:
# средняя квадратическая ошибка
from sklearn.metrics import mean_squared_error
initial_mse =mean_squared_error(y_test, y_pred)
initial_mse

21.517286824088256

In [30]:
# средняя абсолютная ошибка
from sklearn.metrics import mean_absolute_error
mean_absolute_error(y_test, y_pred)

3.1626901513651795

In [31]:
# коэффициент детерминации
from sklearn.metrics import r2_score
r2_score(y_test, y_pred)

0.711228118223995

#### Задание 2.
Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble.
Сделайте агрумент n_estimators равным 1000,
max_depth должен быть равен 12 и random_state сделайте равным 42.
Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression,
но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0],
чтобы получить из датафрейма одномерный массив Numpy,
так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма.
Сделайте предсказание на тестовых данных и посчитайте R2. Сравните с результатом из предыдущего задания.
Напишите в комментариях к коду, какая модель в данном случае работает лучше.


In [32]:
from sklearn.ensemble import RandomForestRegressor

In [33]:
rfr = RandomForestRegressor(n_estimators=1000, max_depth=12,random_state=42)

rfr.fit(X_train, y_train.values[:, 0])
y_pred = rfr.predict(X_test)
y_pred.shape

(152,)

In [34]:
check_test = pd.DataFrame({ 'y_test': y_test['price'], 'y_pred': y_pred.flatten(),})


In [35]:
check_test.head()

y_test     y_pred
173    23.6  22.806412
274    32.4  31.131464
491    13.6  16.339125
72     22.8  23.810726
452    16.1  17.139521

In [36]:
r2_rfr = r2_score(check_test['y_pred'], check_test['y_test'])
r2_rfr

0.8479049999699443

R2_lr=0.7112260057484909 < R2_rfg=0.8479049999699443
R2 больше у модели RandomForestRegressor, значит она лучше объясняет Значения датасета

#### *Задание 3
Вызовите документацию для класса RandomForestRegressor, найдите информацию об атрибуте feature_importances_. С помощью этого атрибута найдите сумму всех показателей важности, установите, какие два признака показывают наибольшую важность.

In [37]:
help(RandomForestRegressor.feature_importances_)

Help on property:

    The impurity-based feature importances.
    
    The higher, the more important the feature.
    The importance of a feature is computed as the (normalized)
    total reduction of the criterion brought by that feature.  It is also
    known as the Gini importance.
    
    high cardinality features (many unique values). See
    :func:`sklearn.inspection.permutation_importance` as an alternative.
    
    Returns
    -------
    feature_importances_ : ndarray of shape (n_features,)
        The values of this array sum to 1, unless all trees are single node
        trees consisting of only the root node, in which case it will be an
        array of zeros.



In [38]:
rfr_imp=rfr.feature_importances_
rfr_imp

array([0.03167574, 0.00154252, 0.00713813, 0.00123624, 0.01426897,
       0.40268179, 0.01429864, 0.06397257, 0.00528122, 0.01152493,
       0.01808108, 0.01245085, 0.41584732])

In [39]:
check_imp = pd.DataFrame({ 'feature_names': feature_names, 'rfr_imp': rfr_imp,})
check_imp

feature_names   rfr_imp
0           CRIM  0.031676
1             ZN  0.001543
2          INDUS  0.007138
3           CHAS  0.001236
4            NOX  0.014269
5             RM  0.402682
6            AGE  0.014299
7            DIS  0.063973
8            RAD  0.005281
9            TAX  0.011525
10       PTRATIO  0.018081
11             B  0.012451
12         LSTAT  0.415847

In [40]:
check_imp.sort_values(by= 'rfr_imp', ascending=False).head()

feature_names   rfr_imp
12         LSTAT  0.415847
5             RM  0.402682
7            DIS  0.063973
0           CRIM  0.031676
10       PTRATIO  0.018081

Очевидно, наиболее важными являются фичи LSTAT и RM, так как имеют наибольшие значениия.